File: FullTextSearch/src/SetOperators/SetOperator.java
package SetOperators;

import Model.InvertedIndex;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

public abstract class SetOperator {
    public Set<String> operate(Set<String> result, final List<String> queryWords, InvertedIndex invertedIndex) {
        Set<String> answer = new HashSet<>(result);
        if (queryWords != null && !queryWords.isEmpty()) {
            for (final String word : queryWords) {
                final Set<String> wordSet = invertedIndex.getIndices().get(word);
                if (wordSet != null && !wordSet.isEmpty()) {
                    specificOperation(answer, wordSet);
                }
            }
        }
        return answer;
    }

    protected abstract void specificOperation(Set<String> result, final Set<String> wordSet);
}

File: FullTextSearchCsharp/FullTextSearch/FullTextSearch/SetOperators/SetOperator.cs
using System.Collections.Generic;
using FullTextSearch.Model;

namespace FullTextSearch.SetOperators
{
    public abstract class SetOperator
    {
        public ISet<string> Operate(ISet<string> result, List<string> queryWords, InvertedIndex invertedIndex)
        {
            ISet<string> answer = new HashSet<string>(result);

            if (queryWords is null) return answer;

            foreach (var word in queryWords)
            {
                if (invertedIndex.Indexes.TryGetValue(word,out var idSet))
                {
                    SpecificOperation(answer, idSet);
                }
            }

            return answer;
        }

        public abstract void SpecificOperation(ISet<string> result, ISet<string> idSet);
    }
}